onnxruntime/contrib_ops/cuda/bert/attention_impl.cu

```diff
@@ -153,408 +153,285 @@ size_t GetAttentionWorkspaceSize(
+  void* value = reinterpret_cast<void*>(data.v);
+  // For packed KV, we can use query input directly.
+  if (data.gemm_buffer == nullptr && data.key != nullptr && data.value == nullptr && data.bias == nullptr) {
+    query = reinterpret_cast<void*>(const_cast<T*>(data.query));
+  }
 
-    DUMP_TENSOR("flash attention output", data.output, batch_size, sequence_length, num_heads, v_head_size);
+  DUMP_TENSOR_INIT();
+  DUMP_TENSOR_D("q(BSNH)", reinterpret_cast<const T*>(query),
+                parameters.batch_size, parameters.sequence_length, parameters.num_heads, parameters.head_size);
+  DUMP_TENSOR_D("k(BSNH)", data.k,
+                parameters.batch_size, parameters.total_sequence_length, parameters.num_heads, parameters.head_size);
+  DUMP_TENSOR_D("v(BSNH)", data.v,
+                parameters.batch_size, parameters.total_sequence_length,
+                parameters.num_heads, parameters.v_head_size);
+
+  ORT_RETURN_IF_ERROR(onnxruntime::flash::mha_fwd(
+      device_prop, stream, query, key, value, data.output, reinterpret_cast<void*>(data.scratch),
+      parameters.batch_size, parameters.num_heads, parameters.num_heads, parameters.head_size,
+      parameters.sequence_length, parameters.total_sequence_length, scale, parameters.is_unidirectional));
+
+  DUMP_TENSOR("flash attention output", data.output,
+              parameters.batch_size, parameters.sequence_length, parameters.num_heads, parameters.v_head_size);
+
+  return Status::OK();
+}
 
-    return Status::OK();
-  }
+template <>
+Status FlashAttention(
+    const cudaDeviceProp& device_prop,
+    cudaStream_t stream,
+    contrib::AttentionParameters& parameters,
+    AttentionData<float>& data,
+    float scale) {
+  return ORT_MAKE_STATUS(ONNXRUNTIME, StatusCode::NOT_IMPLEMENTED, "flash attention does not support float tensor");
+}
 #endif
 
 #if USE_MEMORY_EFFICIENT_ATTENTION
-  if (data.use_memory_efficient_attention) {
-    // We only enable fused cross attention when there is no key padding mask.
-    // Otherwise, key have effective batch size 2 * batch_size, which is different from batch_size of query.
-    assert(qkv.format == AttentionQkvFormat::Q_K_V_BSNH);
-
-    const void* query = qkv.q;
-    const void* key = qkv.k;
-    const void* value = qkv.v;
-    // For packed KV, we can use query input directly.
-    if (data.gemm_buffer == nullptr && data.key != nullptr && data.value == nullptr) {
-      assert(data.bias == nullptr);
-      query = data.query;
-    }
+template <typename T>
+Status EfficientAttention(
+    const cudaDeviceProp& device_prop,
+    cudaStream_t stream,
+    contrib::AttentionParameters& parameters,
+    AttentionData<T>& data,
+    float scale) {
+  // We only enable fused cross attention when there is no key padding mask.
+  // Otherwise, key have effective batch size 2 * batch_size, which is different from batch_size of query.
+  assert(data.qkv_format == AttentionQkvFormat::Q_K_V_BSNH);
+
+  const void* query = data.q;
```

Comment:
maybe a more descriptive name to differentiate data.q from data.query

Reply:
data.query is from node inputs. It could be in Packed QKV format so it might not be the Q that needed in fused attention. 

The q points to actual Q data. I could add some comments later.